resqpy/well/_blocked_well.py

```diff
@@ -575,6 +575,38 @@ def set_for_column(self, well_name, grid, col_ji0, skip_inactive = True, length_
 
         return self.derive_from_dataframe(df, self.well_name, grid, use_face_centres = True, length_uom = length_uom)
 
+    def dtop_dbot_to_layers(self, df, grid):
+        """Populates the wellspec dataframe with corresponding L when DTOP and DBOT are specified but L is not
+        arguments
+          df (pd.DataFrame): Dataframe of wellspec information
+          grid (grid.Grid object): Grid where well is located
+        returns
+          df_expanded (pd.DataFrame): DataFrame of wellspec information where L is populated with the layers corresponding to the DTOP and DBOT depths
+        """
+        corp = grid.corner_points()
+        nk,nj,ni = corp.shape[:3]
+        df_expanded = []
+
+        for ix,data in df.iterrows():
+            if np.isnan(data['L']):
+                dtop = data['DTOP']
+                dbot = data['DBOT']
+                j,i = int(data['JW']), int(data['IW'])
+                for k in range(nk):
+                    if dbot == 'BOT':
+                        data['L'] = k + 1
+                        df_expanded.append(data.copy())
+                    elif (corp[k,j,i,0,:,:,2].min() >= dtop) and (corp[k,j,i,1,:,:,2].max() <= dbot):
```

Comment:
This will include cells that are fully within the dtop to dbot depth range. It would be better if we included cells that are partially in the range, though ideally with a PPERF being set too. Also, it might be slightly better to use the K-, K+ face mid points, rather than the corner points, as the definition of the cell depth range for this purpose.

The corp z units might be increasing upwards (ie. negative values), in which case the z values should be negated before applying this test. You can check the z axis direction in the boolean attribute: grid.crs.z_inc_down which is usually True but not always.